File: ruby/binary_to_decimal_converter/binary_to_decimal_converter_23.rb
# This program is designed to convert binary numbers to their decimal equivalents.
# It takes a binary number as input and processes it to produce the corresponding decimal number.
# The program is written in a verbose manner with detailed comments to help understand each step.

# Function to convert binary to decimal
def binary_to_decimal(binary)
    # Initialize the decimal value to zero
    decimal = 0
    # Initialize the base value to 1, i.e., 2^0
    base = 1
    # Get the length of the binary number
    length = binary.length
    # Loop through each digit of the binary number starting from the end
    (length - 1).downto(0) do |i|
        # If the binary digit is 1, add the base value to the decimal value
        if binary[i] == '1'
            decimal += base
        end
        # Multiply the base value by 2 for the next digit
        base *= 2
    end
    # Return the final decimal value
    decimal
end

# Function to get binary input from the user
def get_binary_input
    # Prompt the user to enter a binary number
    puts "Enter a binary number:"
    # Get the input from the user
    binary = gets.chomp
    # Return the binary input
    binary
end

# Function to display the decimal result
def display_result(decimal)
    # Print the decimal result
    puts "The decimal equivalent is: #{decimal}"
end

# Main function to execute the program
def main
    # Get the binary input from the user
    binary = get_binary_input
    # Convert the binary input to decimal
    decimal = binary_to_decimal(binary)
    # Display the decimal result
    display_result(decimal)
end

# Execute the main function
main

File: javascript/binary_to_decimal_converter/binary_to_decimal_converter_23.js
/**
 * This program is a binary to decimal converter.
 * It is designed to help programmers understand how to convert binary numbers to decimal numbers.
 * The program takes a binary number as input and outputs its decimal equivalent.
 * It includes detailed comments to explain each step of the process.
 */

// Function to convert binary to decimal
function binaryToDecimal(binary) {
    // Initialize the decimal result to 0
    let decimal = 0;
    // Initialize the power of 2 to 1 (2^0)
    let powerOfTwo = 1;

    // Loop through each digit of the binary number starting from the right
    for (let i = binary.length - 1; i >= 0; i--) {
        // Get the current digit (0 or 1)
        let digit = binary[i];

        // If the digit is 1, add the current power of 2 to the decimal result
        if (digit === '1') {
            decimal += powerOfTwo;
        }

        // Multiply the power of 2 by 2 for the next digit
        powerOfTwo *= 2;
    }

    // Return the decimal result
    return decimal;
}

// Function to get user input and display the result
function main() {
    // Prompt the user to enter a binary number
    let binary = prompt("Enter a binary number:");

    // Convert the binary number to decimal
    let decimal = binaryToDecimal(binary);

    // Display the decimal result
    alert("The decimal equivalent is: " + decimal);
}

// Execute the main function
main();

